src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,805 @@
+            private readonly CompletionHelper _completionHelper;
+            private readonly bool _highlightMatchingPortions;
+            private readonly bool _showCompletionItemFilters;
+
+            private readonly Func<ImmutableArray<(RoslynCompletionItem, PatternMatch?)>, string, ImmutableArray<RoslynCompletionItem>> _filterMethod;
+
+            private CompletionTriggerReason InitialTriggerReason => _data.InitialTrigger.Reason;
+            private CompletionTriggerReason UpdateTriggerReason => _data.Trigger.Reason;
+
+            // PERF: Create a singleton to avoid lambda allocation on hot path
+            private static readonly Func<TextSpan, RoslynCompletionItem, Span> s_highlightSpanGetter
+                = (span, item) => span.MoveTo(item.DisplayTextPrefix?.Length ?? 0).ToSpan();
+
+            // We might need to handle large amount of items with import completion enabled,
+            // so use a dedicated pool to minimize/avoid array allocations (especially in LOH)
+            // Set the size of pool to 1 because we don't expect UpdateCompletionListAsync to be
+            // called concurrently, which essentially makes the pooled list a singleton,
+            // but we still use ObjectPool for concurrency handling just to be robust.
+            private static readonly ObjectPool<List<MatchResult<VSCompletionItem>>> s_listOfMatchResultPool = new(factory: () => new(), size: 1);
+
+            public CompletionListUpdater(
+                IAsyncCompletionSession session,
+                AsyncCompletionSessionDataSnapshot data,
+                RecentItemsManager recentItemsManager,
+                IGlobalOptionService globalOptions)
+            {
+                _session = session;
+                _data = data;
+                _recentItemsManager = recentItemsManager;
+
+                _filterText = _session.ApplicableToSpan.GetText(_data.Snapshot);
+
+                if (!_session.Properties.TryGetProperty(CompletionSource.HasSuggestionItemOptions, out bool hasSuggestedItemOptions))
+                {
+                    // This is the scenario when the session is created out of Roslyn, in some other provider, e.g. in Debugger.
+                    // For now, the default hasSuggestedItemOptions is false.
+                    hasSuggestedItemOptions = false;
+                }
+
+                _hasSuggestedItemOptions = hasSuggestedItemOptions || _data.DisplaySuggestionItem;
+
+                // We prefer using the original snapshot, which should always be available from items provided by Roslyn's CompletionSource.
+                // Only use data.Snapshot in the theoretically possible but rare case when all items we are handling are from some non-Roslyn CompletionSource.
+                var snapshotForDocument = TryGetInitialTriggerLocation(_data, out var intialTriggerLocation)
+                    ? intialTriggerLocation.Snapshot
+                    : _data.Snapshot;
```

Comment:
non-roslyn completionsource is a terrifying concept to me :)  this is not on you, but in cases like this we really need to call out what those cases are.  because i genuinely have no idea how we get into this state.

Reply:
You are not alone :)
I'm not certain about this either, therefore reluctant to make change here until understand this better.